File: typescript/src/diff.ts
// JSON diff mirroring Python's diff.py.

export type DiffOp = "changed" | "added" | "removed";
export interface DiffRecord {
  path: string;
  op: DiffOp;
  expected?: unknown;
  actual?: unknown;
}

export function jsonDiff(expected: unknown, actual: unknown, path = ""): DiffRecord[] {
  const out: DiffRecord[] = [];
  walk(expected, actual, path, out);
  return out;
}

function walk(expected: unknown, actual: unknown, path: string, out: DiffRecord[]) {
  if (isObject(expected) && isObject(actual)) {
    const keys = new Set([...Object.keys(expected), ...Object.keys(actual)]);
    for (const k of keys) {
      const sub = path ? `${path}.${k}` : k;
      if (!(k in actual)) out.push({ path: sub, op: "removed", expected: expected[k] });
      else if (!(k in expected)) out.push({ path: sub, op: "added", actual: actual[k] });
      else walk(expected[k], actual[k], sub, out);
    }
    return;
  }
  if (Array.isArray(expected) && Array.isArray(actual)) {
    const n = Math.max(expected.length, actual.length);
    for (let i = 0; i < n; i++) {
      const sub = `${path}[${i}]`;
      if (i >= actual.length) out.push({ path: sub, op: "removed", expected: expected[i] });
      else if (i >= expected.length) out.push({ path: sub, op: "added", actual: actual[i] });
      else walk(expected[i], actual[i], sub, out);
    }
    return;
  }
  if (!deepEqual(expected, actual)) {
    out.push({ path: path || "$", op: "changed", expected, actual });
  }
}

function isObject(v: unknown): v is Record<string, unknown> {
  return v !== null && typeof v === "object" && !Array.isArray(v);
}

function deepEqual(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (typeof a !== typeof b) return false;
  if (a === null || b === null) return a === b;
  if (typeof a === "object") return JSON.stringify(a) === JSON.stringify(b);
  return false;
}

File: python/src/agentreplay/diff.py
"""In-house JSON diff used for divergence reports (see implementation-notes §4.2)."""

from __future__ import annotations

from typing import Any


def json_diff(expected: Any, actual: Any, path: str = "") -> list[dict[str, Any]]:
    """Return a flat list of change records.

    Each record: ``{"path": "a.b[0]", "op": "changed"|"added"|"removed",
    "expected": ..., "actual": ...}``.
    """
    changes: list[dict[str, Any]] = []
    _walk(expected, actual, path, changes)
    return changes


def _walk(expected: Any, actual: Any, path: str, changes: list[dict[str, Any]]) -> None:
    if isinstance(expected, dict) and isinstance(actual, dict):
        for k in expected.keys() | actual.keys():
            sub = f"{path}.{k}" if path else k
            if k not in actual:
                changes.append({"path": sub, "op": "removed", "expected": expected[k]})
            elif k not in expected:
                changes.append({"path": sub, "op": "added", "actual": actual[k]})
            else:
                _walk(expected[k], actual[k], sub, changes)
    elif isinstance(expected, list) and isinstance(actual, list):
        n = max(len(expected), len(actual))
        for i in range(n):
            sub = f"{path}[{i}]"
            if i >= len(actual):
                changes.append({"path": sub, "op": "removed", "expected": expected[i]})
            elif i >= len(expected):
                changes.append({"path": sub, "op": "added", "actual": actual[i]})
            else:
                _walk(expected[i], actual[i], sub, changes)
    else:
        if expected != actual:
            changes.append(
                {"path": path or "$", "op": "changed", "expected": expected, "actual": actual}
            )
